spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatch.java

```diff
@@ -0,0 +1,118 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import java.util.List;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SchemaParser;
+import org.apache.iceberg.SerializableTable;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.spark.SparkReadConf;
+import org.apache.iceberg.spark.source.SparkBatchScan.ReadTask;
+import org.apache.iceberg.spark.source.SparkBatchScan.ReaderFactory;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.connector.read.Batch;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+
+class SparkBatch implements Batch {
+
+  private final JavaSparkContext sparkContext;
+  private final Table table;
+  private final SparkReadConf readConf;
+  private final List<CombinedScanTask> tasks;
+  private final Schema expectedSchema;
+  private final boolean caseSensitive;
+  private final boolean localityEnabled;
+
+  SparkBatch(JavaSparkContext sparkContext, Table table, SparkReadConf readConf,
+             List<CombinedScanTask> tasks, Schema expectedSchema) {
+    this.sparkContext = sparkContext;
+    this.table = table;
+    this.readConf = readConf;
+    this.tasks = tasks;
+    this.expectedSchema = expectedSchema;
+    this.caseSensitive = readConf.caseSensitive();
+    this.localityEnabled = readConf.localityEnabled();
+  }
+
+  @Override
+  public InputPartition[] planInputPartitions() {
+    // broadcast the table metadata as input partitions will be sent to executors
+    Broadcast<Table> tableBroadcast = sparkContext.broadcast(SerializableTable.copyOf(table));
+    String expectedSchemaString = SchemaParser.toJson(expectedSchema);
+
+    InputPartition[] readTasks = new InputPartition[tasks.size()];
+
+    Tasks.range(readTasks.length)
+        .stopOnFailure()
+        .executeWith(localityEnabled ? ThreadPools.getWorkerPool() : null)
+        .run(index -> readTasks[index] = new ReadTask(
+            tasks.get(index), tableBroadcast, expectedSchemaString,
+            caseSensitive, localityEnabled));
+
+    return readTasks;
+  }
+
+  @Override
+  public PartitionReaderFactory createReaderFactory() {
+    return new ReaderFactory(batchSize());
+  }
+
+  private int batchSize() {
+    if (parquetOnly() && parquetBatchReadsEnabled()) {
```

Comment:
I somehow prefer to consider `parquetOnly()` as a part of method `parquetBatchReadsEnabled`. The same to `orcOnly`. So the method is like
```
if(parquetBatchReadsEnabled()){return parquetBatchSize} else if(orcBatchReadsEnabled()) {return orcBatchsize}
```
But it's just my preference, I'm OK with the current one as well.